### ISSUE TITLE
`Decoder::decode()` says an Err informs `Framed` that the stream should be terminated, but `Framed` doesn't terminate

### ISSUE THREAD

**Issue by lilyball** [tokio-rs/tokio#3976]
**Version**
`tokio-util v0.6.7`

**Platform**
Darwin My-Computer 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64 x86_64

**Description**
`tokio_util::codec::FramedRead` will happily continue trying to decode a stream after the associated `tokio_util::codec::Decoder` implementation returns `Err` from `Decoder::decode()` or `Decoder::decode_eof()`. This happens in spite of the documentation on `Decoder::decode()`:

> Finally, if the bytes in the buffer are malformed then an error is returned indicating why. This informs `Framed` that the stream is now corrupt and should be terminated.

(`decode_eof()` does not document the semantics of returning an error but it's safe to assume it's the same as `decode()`).

Looking at `<FramedImpl as Stream>::poll_next()`, it returns the `Err` to its caller without modifying its state:

https://github.com/tokio-rs/tokio/blob/2087f3e0ebb08d633d59c5f964b3901e68b3c038/tokio-util/src/codec/framed_impl.rs#L168-L179

This means that if the `decode()` method returns an `Err` without consuming any bytes from the buffer, a simple loop like the following will go into a tight infinite loop:

```rust
while let Some(result) = framed.next().await {
    process(result);
}
```

And if a `Decoder` impl can return recoverable errors, and does so from the `decode()` method instead of moving them into the `Item`, then that lends itself to the following loop:

```rust
while let Some(result) = framed.next().await {
    match result {
        Ok(val) => process(val),
        Err(e) => warn!(log, "error decoding stream"; "error" => %e),
    }
}
```

It's really easy to write this loop, especially if you haven't carefully read all of the associated documentation, and you won't realize what happened until you hit a nonrecoverable error.

If `Framed` actually did what the `Decoder::decode()` documentation suggests and moved to a terminal state upon encountering an `Err`, then these loops would instantly go from an infinite loop to one that logs an error and terminates. This would also strongly encourage `Decoder` impls that can return recoverable errors to make that distinction properly in its return type.

Optionally, if someone is using a non-conforming `Decoder` impl that mixes recoverable and non-recoverable errors together and the calling code wants to intelligently handle that, `Framed` could offer a method to recover from the terminal state.

This behavior change should probably be considered a breaking change, but I could make the argument that it's a bugfix (makes behavior conform to documentation) and that it's an important safety improvement for existing code (turns infinite loops into finite ones).

**Comment by Darksonn** [comment]
It seems reasonable that it should return `None` after an error.

**Comment by bIgBV** [comment]
@Darksonn I can take a look at this issue.

**Comment by Darksonn** [comment]
Fixed in #4166
